File: app-laravel/app/Http/Controllers/FailedJobController.php
<?php

declare(strict_types=1);

namespace App\Http\Controllers;

use App\Http\Resources\ApiResponse;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Artisan;
use Illuminate\Support\Facades\DB;

/**
 * FAILED JOB CONTROLLER
 * =====================
 * Uğursuz queue job-larını idarə etmək üçün API.
 *
 * QUEUE FAILED JOBS NƏDİR?
 * Job bütün retry cəhdlərindən sonra hələ uğursuz olursa,
 * Laravel onu failed_jobs cədvəlinə yazır.
 *
 * Bu controller admin-lərə imkan verir:
 * - Uğursuz job-ları görmək
 * - Yenidən cəhd etmək (retry)
 * - Silmək
 *
 * REAL DÜNYADA:
 * - Admin dashboard-da "Failed Jobs" bölməsi olur
 * - Monitoring tool (Horizon, Telescope) bunu vizual göstərir
 * - Alert sistemi uğursuz job-lar haqqında bildiriş göndərir
 */
class FailedJobController extends Controller
{
    /**
     * GET /api/admin/failed-jobs
     * Bütün uğursuz job-ların siyahısı.
     */
    public function index(): JsonResponse
    {
        $failedJobs = DB::table('failed_jobs')
            ->orderByDesc('failed_at')
            ->paginate(20);

        return ApiResponse::success(data: $failedJobs);
    }

    /**
     * GET /api/admin/failed-jobs/{id}
     * Tək uğursuz job-un detalları.
     */
    public function show(string $id): JsonResponse
    {
        $job = DB::table('failed_jobs')->where('uuid', $id)->first();

        if (!$job) {
            return ApiResponse::error('Uğursuz job tapılmadı', code: 404);
        }

        return ApiResponse::success(data: $job);
    }

    /**
     * POST /api/admin/failed-jobs/{id}/retry
     * Uğursuz job-u yenidən queue-ya göndər.
     *
     * queue:retry artisan command-ı istifadə edir.
     * Job failed_jobs-dan silinir və orijinal queue-ya qayıdır.
     */
    public function retry(string $id): JsonResponse
    {
        $job = DB::table('failed_jobs')->where('uuid', $id)->first();

        if (!$job) {
            return ApiResponse::error('Uğursuz job tapılmadı', code: 404);
        }

        Artisan::call('queue:retry', ['id' => [$id]]);

        return ApiResponse::success(message: 'Job yenidən queue-ya göndərildi');
    }

    /**
     * POST /api/admin/failed-jobs/retry-all
     * Bütün uğursuz job-ları yenidən cəhd et.
     */
    public function retryAll(): JsonResponse
    {
        $count = DB::table('failed_jobs')->count();

        if ($count === 0) {
            return ApiResponse::success(message: 'Uğursuz job yoxdur');
        }

        Artisan::call('queue:retry', ['id' => ['all']]);

        return ApiResponse::success(
            message: "{$count} job yenidən queue-ya göndərildi"
        );
    }

    /**
     * DELETE /api/admin/failed-jobs/{id}
     * Uğursuz job-u sil.
     */
    public function destroy(string $id): JsonResponse
    {
        $deleted = DB::table('failed_jobs')->where('uuid', $id)->delete();

        if ($deleted === 0) {
            return ApiResponse::error('Uğursuz job tapılmadı', code: 404);
        }

        return ApiResponse::success(message: 'Uğursuz job silindi');
    }

    /**
     * DELETE /api/admin/failed-jobs
     * Bütün uğursuz job-ları sil.
     */
    public function flush(): JsonResponse
    {
        $count = DB::table('failed_jobs')->count();
        Artisan::call('queue:flush');

        return ApiResponse::success(
            message: "{$count} uğursuz job silindi"
        );
    }
}

File: app/spring/src/main/java/az/ecommerce/admin/FailedJobController.java
package az.ecommerce.admin;

import az.ecommerce.shared.infrastructure.api.ApiResponse;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Laravel: app/Http/Controllers/FailedJobController.php
 *   - Failed job-ları idarə edir (DLQ)
 *
 * Spring: RabbitMQ DLQ (dead-letter exchange) → dead_letter_messages cədvəli
 * Real implementasiya RabbitAdmin və DLQ-dən oxuma tələb edir.
 */
@RestController
@RequestMapping("/api/admin/failed-jobs")
@PreAuthorize("hasRole('ADMIN')")
public class FailedJobController {

    @GetMapping
    public ApiResponse<List<Map<String, Object>>> index() {
        // dead_letter_messages cədvəlindən oxuma
        return ApiResponse.success(List.of());
    }

    @GetMapping("/{id}")
    public ApiResponse<Map<String, Object>> show(@PathVariable UUID id) {
        return ApiResponse.success(Map.of("id", id, "status", "failed"));
    }

    @PostMapping("/{id}/retry")
    public ApiResponse<Void> retry(@PathVariable UUID id) {
        // Republish to original queue
        return ApiResponse.success(null, "Yenidən cəhd edildi");
    }

    @PostMapping("/retry-all")
    public ApiResponse<Void> retryAll() {
        return ApiResponse.success(null, "Bütün failed job-lar yenidən cəhd edildi");
    }

    @DeleteMapping("/{id}")
    public ApiResponse<Void> destroy(@PathVariable UUID id) {
        return ApiResponse.success(null, "Silindi");
    }

    @DeleteMapping
    public ApiResponse<Void> flush() {
        return ApiResponse.success(null, "Bütün failed job-lar silindi");
    }
}
